### DISCUSSION TITLE
Password-less authentication

### DISCUSSION THREAD

**Discussion by accuser** [swe-delivery/swe-tube#5] [Ideas]
For users to make contributions, vote on contributions by others, and discuss contributions, we will need some form of authentication. 

I've got a password-less authentication service implemented as a few serverless functions that I could repurpose for this.

Essentially, when a user signs up they provide an email address which is then confirmed with a registration token (JWT) to their inbox allowing them to be onboarded. Successful registration generates an identity token and an access token. If these expire (i.e., not refreshed due to inactivity) then the next time the user accesses the app they are asked for their email address, to which an authentication token is sent allowing them to be authenticated. The identify token provides user displayable information for the front-end (by way of claims) while the access token is used for role-based access control in the API.

We could allow self-registration, but I suspect that a curated onboarding (i.e., by invitation) is probably more suitable for our context.

**Comment by nimonian** [comment]
Since we're going to build a "bootcamp" version followed by a "kitchen-sink" fork, is this something we could add to the latter?

I'm cautious of making the bootcamp version too sophisticated and even willing to make a few egregious imperfections to keep it simple and within their grokability

Could it be anonymous for the bootcamp version? Maybe save a visitor id in local storage? I think the bc version should try to avoid auth altogether as silly as that sounds

I user https://colorhunt.co/ quite a bit and I think it does basically that

**Reply by accuser** [reply]
If this is going to be a live service - which I am assuming that it is as I think we can all see a need for it - then I think we do need some authentication. We don't need anything for accessing the site (i.e., read-only), but for contributions / submissions we will need to know who is accessing it.

Looking at https://colorhunt.co/ they are tracking users with a number of ad services, so they don't really mind, and they don't accept submissions, so there is no need to moderate offensive colour palettes. 

We could generate a unique visitor id based on a fingerprint of the User Agent so that likes could be tracked, but that doesn't work for users accessing the site with both mobile and desktop browsers.

But having a vanilla service and one with sprinkles is a good challenge.

**Reply by nimonian** [reply]
Moderating offensive content definitely becomes an issue with anonymous submission - we don't want a chan board

This in an important discussion to have actually because loads of our bootcamp projects are anonymous CRUD apps so apprentices will be in the same situation

Imo it's best if the vanilla version should be something bootcampers could build knowing only bootcamp stuff, which means we don't have auth. Is it reasonable to simply never deploy vanilla? It's an _exercise_ with the huge caveat that it should never hit the open internet. It should be dead simple - too simple to survive in the wild. An incomplete version.

We could then branch or fork and refactor to incorporate auth & roles: a deployable version.

> but that doesn't work for users accessing the site with both mobile and desktop browsers.

Definitely - I think this is ok tbh, it's a deficiency baked right in to the fact that we don't teach auth concepts during bootcamp. Imagine they're just making a todo app - they wouldn't be able to open the same todos on mobile and browser, and somebody could create highly offensive todos! That's fine - it's about CRUDing todos in the database and getting the stack working together.

**Reply by nimonian** [reply]
By reaching a conclusion here we're answering the key question: _what if a bootcamper decided to build this project? what would we tell them to do?_ Really important I think!

**Reply by Jaardy** [reply]
Could we fake a launch and then have to shut it down because of this concern?
It would actually roll really well into how I've approached the post bootcamp modules.  I've suggested that we are redoing the bootcamp but with an eye to the professional context.  This could be the "event" that shifts our assumptions around the intended audience and their intent.

**Reply by Jaardy** [reply]
> no need to moderate offensive colour palettes.

I would be banned for this 😆

**Reply by accuser** [reply]
We might not teach auth on bootcamp, but just about every end-of-bootcamp project (SEP22) included auth. It was like a defining moment for those that implemented it. Lots of them came and showed me. 😄

**Reply by nimonian** [reply]
I know the vanilla version is unsophisticated, unrealistic even, but it's more about making a UI, API and data layer work together in a way that we can extrapolate back from to inform how we communicate and structure projects in weeks 4 - 9 of bootcamp. I do feel that an authless CRUD is the most useful model for this purpose.

Perhaps doing something social without auth is too silly - so maybe we strip out the social aspect for the vanilla version? It's a place to keep track of, describe, tag and "mark complete" your _own_ resources - a beefed up bookmark bar. This is much closer to the type of project we ask them to build in the bootcamp.

**Reply by accuser** [reply]
So I'm thinking that we should probably build the vanilla version as something that an apprentice could spin up locally but maintaining some utility. How about we build a service for resources separately (which is exposed via a REST API) that we can populate? The vanilla project then becomes a consumer of that API, being able to fetch resources, filtered and paginated. The vanilla project includes a local persistence to track their own tags, reflections, etc., so consuming an API and adding value. And as instructors we can still gather and curate resources that we think are valuable.

**Reply by accuser** [reply]
We can still include the `api` in the project repo, but coming from bootcamp all they need to be concerned about is the `web` app. Having it there for discovery would be instructional.

**Reply by nimonian** [reply]
This sounds perfect - we can demonstrate all the coding standards and organisation that we'd like, and its easily extensible into a useful product. Great solution.

**Comment by accuser** [comment]
Another thought - but park this - what about simply adding an `access_token` in the `authorization` header of the API `fetch` request? This is something that they might be familiar with (I know that some in the last bootcamp used a weather API that required this) and each apprentice could get their own `access_token`. No need for authentication.

**Reply by nimonian** [reply]
This is a really neat idea. Do you imagine using this strategy for the sprinkles version too? Or would we use this only for the vanilla version and have an authentication strategy for sprinkles?
